e2etests/tests/status.go

```diff
@@ -349,6 +350,95 @@ var _ = ginkgo.Describe("Exposing FRR status", func() {
 			}, 30*time.Second, time.Second).ShouldNot(HaveOccurred())
 		})
 	})
+
+	ginkgo.Context("Node state cleaner functionality", func() {
+		const podNameAnnotation = "frrk8s.metallb.io/pod-name"
+
+		ginkgo.It("should delete FRRNodeState when corresponding pod is deleted", func() {
+			nodes, err := k8s.Nodes(cs)
+			Expect(err).NotTo(HaveOccurred())
+
+			ginkgo.By("Verifying all node states exist with pod annotations")
+			statuses := frrk8sv1beta1.FRRNodeStateList{}
+			err = cl.List(context.Background(), &statuses)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(len(statuses.Items)).To(Equal(len(nodes)))
+
+			for _, status := range statuses.Items {
+				Expect(status.Annotations).NotTo(BeNil())
+				Expect(status.Annotations[podNameAnnotation]).NotTo(BeEmpty())
+			}
+
+			ginkgo.By(fmt.Sprintf("Finding frr-k8s pod on first node %s to delete", nodes[0].Name))
+			podToDelete, err := k8s.FRRK8sPodOnNode(cs, nodes[0].Name)
+			Expect(err).NotTo(HaveOccurred())
+			ginkgo.By(fmt.Sprintf("Deleting pod %s", podToDelete.Name))
+
+			err = cs.CoreV1().Pods(k8s.FRRK8sNamespace).Delete(context.Background(), podToDelete.Name, metav1.DeleteOptions{})
+			Expect(err).NotTo(HaveOccurred())
+
+			Eventually(func() bool {
+				_, err := cs.CoreV1().Pods(k8s.FRRK8sNamespace).Get(context.Background(), podToDelete.Name, metav1.GetOptions{})
+				return errors.IsNotFound(err)
+			}, time.Minute, time.Second).Should(BeTrue())
+
+			ginkgo.By("Waiting for FRRNodeState to be deleted")
```

Comment:
I might be missing something, but this test is possibly a vw? (since we don't really change the daemonset's nodeSelector)
we're checking if a state was deleted by going over all of the states and checking that none contains the old pod name, but it could be just that the new pod overwrote the previous state, so this test would pass even without the cleaner controller running

Reply:
yeah I figured, I guess a fair compromise can be manually creating a nodestate pointing to a non-existing pod / with a bad node name (depending on the other discussion), and seeing it get deleted